packages/repository/test/unit/repositories/constraint-utils.unit.ts

```diff
@@ -0,0 +1,113 @@
+// Copyright IBM Corp. 2017,2018. All Rights Reserved.
+// Node module: @loopback/repository
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {expect} from '@loopback/testlab';
+import {
+  FilterBuilder,
+  Filter,
+  Fields,
+  Where,
+  WhereBuilder,
+} from '../../../src/query';
+import {AnyObject} from 'loopback-datasource-juggler';
+import {
+  constrainFilter,
+  constrainWhere,
+} from '../../../src/repositories/constraint-utils';
+
+describe('constraint utility functions', () => {
+  let originalFilter: Filter = {};
+  let originalWhere: Where = {};
+  const constraint: AnyObject = givenAWhereConstraint();
+
+  before(() => {
+    originalFilter = givenAFilter({a: true}, {x: 'x'}, 5);
+    originalWhere = givenAWhere();
+  });
+  context('constrainFilter', () => {
+    it('applies a where constraint', () => {
+      const result = constrainFilter(originalFilter, constraint);
+      expect(result).to.deepEqual({
+        fields: {a: true},
+        limit: 5,
+        where: {x: 'x', id: '5'},
+      });
+    });
+    it('applies a filter constraint with where object', () => {
+      const validFilterConstraint = givenAFilter(
+        undefined,
+        {id: '10'},
+        undefined,
+      );
```

Comment:
Positional arguments are difficult to understand in this case - what is `undefined` standing for?

Please consider using a single argument containing an object with named properties.

```ts
givenAFilter({fields: undefined, where: {id: '10'}, limit: undefined});
// I think this can be further simplified to
givenAFilter({where: {id: '10'}});
```

Do we actually need to pass that data object through `givenAFilter`, considering that the input is already a valid filter and most likely the same value as will be returned by `givenAFilter`?

Reply:
I agree that using `givenAFilter` function is moot. I've removed it.